### PR TITLE
Add templates for presentation viewer

### PR DESCRIPTION
**Pull request by PrajwolAmatya**
## Description
This PR adds basic templates for the presentation viewer.

Following templates are added:
1. Cover Slide Template

image

2. Title Content Slide Template

image

3. Title Content Image Slide Template

image

4. About Us Slide Template

image

### Slide content fitting

The font size of the slide will be automatically reduced when the content tends to overflow out of the slide.

image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -50,6 +50,208 @@ This app has the following default slide separators:
...
+| `logo`           | Logo of the company. Logo appears on the top right corner of the slide                                                                                     |
+| `color`          | Font color of the slide title.                                                                                                                             |
+| `footer`         | Content to be shown in the footer of the slide.                                                                                                            |
+| `aboutUs`        | Column wise content to be shown in the about-us slide. Use this metadata only if about-us slide is used for the presentation. More [here](#slide-about-us) |
+
+Following the front matter, create slides as described in [Creating Presentation](#creating-presentation)
+
+Besides default metadata, you can also provide inline metadata for each slide. This will be useful when you want to
+override the default metadata for a specific slide and this will be applied to that slide only. The inline metadata can
+be defined by adding metadata directly after the slide title as below:
```
Comment (saw-jan) [inline README.md]:
```suggestion
be defined by adding metadata directly on the slide title as below:
```

File: src/App.vue [hunk 2]
```diff
@@ -207,6 +233,172 @@ function dirname(path: string) {
...
+  for (const [slideNumber, slide] of Array.from(allSlides).entries()) {
+    const customSlideNumber = slide.querySelector('.custom-slide-number')
+    if (!customSlideNumber) {
+      continue
+    }
+    customSlideNumber.textContent = `${slideNumber + 1}`
+  }
+}
+function updateImageStructure() {
+  const pTags = document.querySelectorAll('p > img')
```
Comment (saw-jan) [inline src/App.vue]: element selection without scope grabs the elements globally

File: src/App.vue [hunk 3]
```diff
@@ -207,6 +233,174 @@ function dirname(path: string) {
...
+        const padding = parseFloat(style.padding) * scaleFactor
+        box.style.padding = `${Math.max(2, Math.floor(padding))}px`
+        box.style.margin = `${Math.max(2, Math.floor(parseFloat(style.margin) * scaleFactor))}px`
+      })
+      infoSectionWidth = infoSection.scrollWidth
+    }
+  }
+}
+function fitContent() {
+  const images = document.querySelectorAll('img')
```
Comment (saw-jan) [inline src/App.vue]: here as well. let's scope the selector when using `document.`

File: README.md [hunk 4]
```diff
@@ -50,6 +50,215 @@ This app has the following default slide separators:
...
+
+The slide template should be provided in the default metadata. Provide slide template that will be mostly used in the
+default metadata, so you don't need to add the slide metadata in every slide.
+
+| Slide Template        | Desctiption                                                     | Usage                         |
+|-----------------------|-----------------------------------------------------------------|-------------------------------|
+| `cover`               | This slide template can be used for cover slide                 | `::slide:cover`               |
+| `title-content`       | The slide with title and content (content can be image as well) | `::slide:title-content`       |
+| `title-content-image` | The slide with title, and content in left and image in right    | `::slide:title-content-image` |
+| `about-us`            | The slide with company info (e.g., Values, mission, Vision)     | `::slide:about-us`            |
```
Comment (saw-jan) [inline README.md]:
```suggestion
| `about-us`            | The slide with about-us info (e.g., Values, mission, Vision)     | `::slide:about-us`            |
```

File: README.md [hunk 5]
```diff
@@ -50,6 +50,215 @@ This app has the following default slide separators:
...
+default metadata can be defined as below:
+
+```markdown
+---
+slide: <default-slide-type>
+presenter: <name-of-the-presenter>
+logo: <path-to-logo>
+color: <font-color-for-slide-title>
+footer: <footer-content>
+aboutUs: <content-for-about-us-slide>
```
Comment (saw-jan) [inline README.md]: where do I see this content. I do not see in the about-us slide
```md
---
slide: cover
presenter: Tester
logo: ./car.png
color: red
footer: my foot
aboutUs: this is about us
---

## Slide 1 ::slide:cover

content one

---
## Slide 2 ::slide:title-content

content two

---
## About Us ::slide:about-us

content two
```
Screenshot from 2025-10-31 10-45-09

File: README.md [hunk 6]
```diff
@@ -50,6 +50,215 @@ This app has the following default slide separators:
...
+
+To use the templates in the presentation viewer, use the front matter to provide default metadata followed by slide
+contents.
+
+The front matter is used to define the default values for the presentation and will be applied to every slides. The
+default metadata can be defined as below:
+
+```markdown
+---
+slide: <default-slide-type>
```
[the remark this answers is not in the document]
Comment (PrajwolAmatya) [reply README.md]: The slide variable is already defined in the plugin.

File: README.md (hunk 5 above)
Comment (PrajwolAmatya) [reply README.md]: Missed to update the documenatation here. it should be as below
```markdown
---
slide: <default-slide-type>
presenter: <name-of-the-presenter>
logo: <path-to-logo>
color: <font-color-for-slide-title>
footer: <footer-content>
aboutUs:
    - title: <about-us-title1>
      text: <text-for-title1>
    - title: <about-us-title2>
      text: <text-for-title2>
    - title: <about-us-title3>
      text: <text-for-title3>
---
```

the reason for doing this is we could not split the about-us contents into different columns in the templates using mustache.

In `revealjs-awesoMD` plugin, we send the content to the slide as
```js
tempDiv.innerHTML = Mustache.render(xhr.responseText, {
    title: title,
    content: slideContent,
    metadata: options.metadata,
})
```
Here title is the title of the slide and we send the content through the content variable.

For the following markdown:
```markdown
# Title of slide
hello world
- some
- content
```
here,
title = Title of slide
content = 
```
hello world
- some
- content
```

In the templates we have
```html
<div class="content">
<div class="about-us-text">

{{{ content }}}

</div>
<div class="info-section">

{{#metadata.aboutUs}}
<div class="info-box">
<h3>{{ title }}</h3>
<p>{{ text }}</p>
</div>
<div class="divider"></div>
{{/metadata.aboutUs}}

</div>

</div>
```
Here if we provide the column wise contents in the content variable, we cannot split the content column wise as the content to show in each column could be anything, which makes it difficult to differentiate between the content.
That's why we provide these data from the default metadata in the frontmatter.

**Review by saw-jan — APPROVED** [review]
👍 ✨

File: README.md [hunk 7]
```diff
@@ -50,6 +50,223 @@ This app has the following default slide separators:
...
+```markdown
+# Title of slide ::slide:title-content-image ::logo:logo.png
+```
+
+For this slide, the `slide` template will be set to `title-content-image`, and the `logo` will be set to `logo.png`.
+
+### Slide Templates
+
+The slide template should be provided in the default metadata. Provide slide template that will be mostly used in the
+default metadata, so you don't need to add the slide metadata in every slide.
```
Comment (individual-it) [inline README.md]: This is confusing, I want to  have different templates for different slides. But the template that is mostly used should be set in the default. Is that what you want to say?

**Review by individual-it — CHANGES_REQUESTED** [review]

**Comment by individual-it** [comment]
pnpm update and then make command did not help

File: README.md [hunk 8]
```diff
@@ -50,6 +50,222 @@ This app has the following default slide separators:
...
+
+To use the templates in the presentation viewer, use the front matter to provide default metadata followed by slide
+contents.
+
+The front matter is used to define the default values for the presentation and will be applied to every slides. The
+default metadata can be defined as below:
+
+```markdown
+---
+slide: <default-slide-type>
```
[the remark this answers is not in the document]
Comment (individual-it) [reply README.md]: not giving a default slide, breaks the presentation

**Review by individual-it — CHANGES_REQUESTED** [review]
would be good to have a "blank" template that works just as the default reveal-js template

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (PrajwolAmatya) [reply README.md]: > not giving a default slide, breaks the presentation

That was because i only checked if slide is provided in the default metadata to load styling for the templates. I have fixed it.

> special characters in the slide-name e.g. `<title-content>` or `<title-content` break the presentation

It is because the slide type doesn't exist. So, it should not break the whole presentation. Just the slide where invalid slide type is defined, It currently shows blank slide.
We are supposed to provide valid slide name. It should work same as what we did in openproject. The slide name is actually checked in the plugin.

[the remark this answers is not in the document]
Comment (individual-it) [reply README.md]: If I provide `slide: title-content-does-not-exist` in the default metadata I get

grafik

:+1:

If I provide: `slide: <title-content>` I get
grafik

Not sure what happens but would be good to see the same error message

File: README.md [hunk 9]
```diff
@@ -50,6 +50,222 @@ This app has the following default slide separators:
...
+presenter: <name-of-the-presenter>
+logo: <path-to-logo>
+color: <font-color-for-slide-title>
+footer: <footer-content>
+aboutUs:
+    - title: <about-us-title1>
+      text: <text-for-title1>
```
Comment (individual-it) [inline README.md]: how difficult would it be to allow links here?
Currently
```
  - title: WWW
     text: [www.jankaritech.com](www.jankaritech.com)
```

makes the whole presentation to fall apart

```
Object { name: "YAMLException", reason: "bad indentation of a mapping entry", mark: {…}, message: "bad indentation of a mapping entry at line 7, column 10:\n         text: [www.jankaritech.com](www.jank ... \n             ^", stack: "", … }
​
mark: Object { buffer: "slide: title-content\npresenter:  JankariTech Pvt. Ltd. \nlogo: Logo-colour.png\nfooter: Markdown Presentation Viewer\naboutUs:\n  - title: WWW\n     text: [www.jankaritech.com](www.jankaritech.com)\n  - title: linkedin\n    text: www.linkedin.com/company/jankaritech\n  - title: facebook\n    text: www.facebook.com/jankaritech\n\u0000", position: 148, line: 6, … }
​
message: "bad indentation of a mapping entry at line 7, column 10:\n         text: [www.jankaritech.com](www.jank ... \n             ^"
​
name: "YAMLException"
​
reason: "bad indentation of a mapping entry"
```

If links are not supported that is fine, but the presentation should not break

**Comment by PrajwolAmatya** [comment]
I have addressed the issues and fixed them. Most of the fixes needed to done in the plugin (PR: https://github.com/opf/revealjs-awesoMD/pull/24). 


> Let's add tests for the templating to make sure all these corner cases don't break

Will work on it. :+1: 

@individual-it

**Review by individual-it — CHANGES_REQUESTED** [review]
anything shown to the user e.g. error messages should be sanitized.

currently I can do:
```
# This template does not exist and has special char  ::slide:<a href="jankari.tech">jankaritech</a>
```

and the browser will render `<a href="jankari.tech">jankaritech</a>` as HTML

**Review by individual-it — CHANGES_REQUESTED** [review]
sometimes the logo works sometimes not

I am using
```
---
slide: title-content
presenter:  JankariTech Pvt. Ltd. 
logo: images/Logo-colour.png
```
and the file `images/Logo-colour.png` does exist.

the URL of the logo in the presentation is sometimes: https://localhost:9200/images/Logo-colour.png but after a reload correctly `blob:https://localhost:9200/9db3f52b-2d67-4be4-b116-9a94f6798c7`

**Comment by PrajwolAmatya** [comment]
> anything shown to the user e.g. error messages should be sanitized.
> 
> currently I can do:
> 
> ```
> # This template does not exist and has special char  ::slide:<a href="jankari.tech">jankaritech</a>
> ```
> 
> and the browser will render `<a href="jankari.tech">jankaritech</a>` as HTML

Fixed it in PR https://github.com/opf/revealjs-awesoMD/pull/25.
Screenshot from 2025-11-20 10-59-52

**Review by individual-it — APPROVED** [review]
